specs/_features/whisk/beacon-chain.md

```diff
@@ -0,0 +1,490 @@
+```
+
+### Whisk
+
+#### `BeaconBlockBody`
+
+```python
+class BeaconBlockBody(capella.BeaconBlockBody):
+    randao_reveal: BLSSignature
+    eth1_data: Eth1Data  # Eth1 data vote
+    graffiti: Bytes32  # Arbitrary data
+    # Operations
+    proposer_slashings: List[ProposerSlashing, MAX_PROPOSER_SLASHINGS]
+    attester_slashings: List[AttesterSlashing, MAX_ATTESTER_SLASHINGS]
+    attestations: List[Attestation, MAX_ATTESTATIONS]
+    deposits: List[Deposit, MAX_DEPOSITS]
+    voluntary_exits: List[SignedVoluntaryExit, MAX_VOLUNTARY_EXITS]
+    sync_aggregate: SyncAggregate
+    # Execution
+    execution_payload: ExecutionPayload
+    # Capella operations
+    bls_to_execution_changes: List[SignedBLSToExecutionChange, MAX_BLS_TO_EXECUTION_CHANGES]
+    whisk_post_shuffle_trackers: Vector[WhiskTracker, WHISK_VALIDATORS_PER_SHUFFLE]  # [New in Whisk]
+    whisk_shuffle_proof: WhiskShuffleProof  # [New in Whisk]
+    whisk_shuffle_proof_M_commitment: BLSG1Point  # [New in Whisk]
+    whisk_registration_proof: WhiskTrackerProof  # [New in Whisk]
+    whisk_tracker: WhiskTracker  # [New in Whisk]
+    whisk_k_commitment: BLSG1Point  # [New in Whisk]
+```
+
+```python
+def get_shuffle_indices(randao_reveal: BLSSignature) -> Sequence[uint64]:
+    """
+    Given a `randao_reveal` return the list of indices that got shuffled from the entire candidate set
+    """
+    indices = []
+    for i in WHISK_VALIDATORS_PER_SHUFFLE:
+        # XXX ensure we are not suffering from modulo bias
+        shuffle_index = uint256(hash(randao_reveal + uint_to_bytes(i))) % WHISK_CANDIDATE_TRACKERS_COUNT
+        indices.append(shuffle_index)
+
+    return indices
+```
+
+```python
+def whisk_process_shuffled_trackers(state: BeaconState, body: BeaconBlockBody) -> None:
+    # Check the shuffle proof
+    shuffle_indices = get_shuffle_indices(body.randao_reveal)
+    pre_shuffle_trackers = [state.whisk_candidate_trackers[i] for i in shuffle_indices]
+    post_shuffle_trackers = body.whisk_post_shuffle_trackers
+
+    shuffle_epoch = get_current_epoch(state) % WHISK_EPOCHS_PER_SHUFFLING_PHASE
+    if shuffle_epoch + WHISK_PROPOSER_SELECTION_GAP + 1 >= WHISK_EPOCHS_PER_SHUFFLING_PHASE:
+        # Require unchanged trackers during cooldown
+        assert pre_shuffle_trackers == post_shuffle_trackers
+    else:
+        # Require shuffled trackers during shuffle
+        assert IsValidWhiskShuffleProof(
+            pre_shuffle_trackers,
+            post_shuffle_trackers,
+            body.whisk_shuffle_proof_M_commitment,
+            body.whisk_shuffle_proof,
+        )
+
+    # Shuffle candidate trackers
+    for i, shuffle_index in enumerate(shuffle_indices):
+        state.whisk_candidate_trackers[shuffle_index] = post_shuffle_trackers[i]
+```
+
+```python
+def is_k_commitment_unique(state: BeaconState, k_commitment: BLSG1Point) -> bool:
+    return all([validator.whisk_k_commitment != k_commitment for validator in state.validators])
```

Comment:
scary due to its syntax, efficiency or else? If the former should I use a for loop?

Reply:
Agreed. Let's use a for loop?

Reply:
Apologies for the confusion. I was scared by the efficiency. IIUC Python `all()` is like for loop implementation, it returns `False` immediately once it finds `validator.whisk_k_commitment == k_commitment`. But it scares me as it is used in the `while True` of `get_unique_whisk_k`.

Reply:
Regarding efficiency: Yes on a deposit this requires a full loop over the validator list. But a naive implementation of the beacon chain already requires that to check if the pubkey of a deposit has been seen before or not. This addition will require clients to mantain cached set of k_commitment to avoid that loop if they choose. AFAIK spec should not be concerned with computational efficiency but correctness, and the current whisk spec new k_commitment to be unique. 

Regarding syntax: the all syntax is pretty readable, I can switch to a for loop, but the current implementation is simpler IMO